ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java

```diff
@@ -646,41 +680,72 @@ public Page<String> listAllValidDatasetVersions(short accountId, short container
    * @param datasetName the name of the dataset.
    * @return a list of {@link DatasetVersionRecord}
    */
-  public List<DatasetVersionRecord> getAllValidVersionsOutOfRetentionCount(short accountId,
-      short containerId, String accountName, String containerName, String datasetName)
-      throws SQLException, AccountServiceException {
+  public List<DatasetVersionRecord> getAllValidVersionsOutOfRetentionCount(short accountId, short containerId,
+      String accountName, String containerName, String datasetName) throws SQLException, AccountServiceException {
     long startTimeMs = System.currentTimeMillis();
     Integer retentionCount;
     Dataset.VersionSchema versionSchema;
+    String retentionPolicy;
+    List<DatasetVersionRecord> datasetVersionRecordList;
     try {
       //if dataset is deleted, we should not be able to get any dataset version.
       Dataset dataset = getDataset(accountId, containerId, accountName, containerName, datasetName);
       retentionCount = dataset.getRetentionCount();
       versionSchema = dataset.getVersionSchema();
-    } catch (SQLException | AccountServiceException e) {
-      dataAccessor.onException(e, Read);
-      throw e;
-    }
-    if (retentionCount != null) {
-      try {
-        dataAccessor.getDatabaseConnection(false);
-        PreparedStatement listAllValidVersionsOrderedByLastModifiedTimeStatement =
-            dataAccessor.getPreparedStatement(listVersionByModifiedTimeAndFilterByRetentionSql, false);
-        List<DatasetVersionRecord> datasetVersionRecordList =
-            executeListAllValidVersionsOrderedByLastModifiedTimeStatement(
-                listAllValidVersionsOrderedByLastModifiedTimeStatement, accountId, containerId, datasetName,
-                retentionCount, versionSchema);
+      retentionPolicy = dataset.getRetentionPolicy();
+      datasetVersionRecordList = new ArrayList<>();
+      if (retentionCount == null) {
         dataAccessor.onSuccess(Read, System.currentTimeMillis() - startTimeMs);
         return datasetVersionRecordList;
-      } catch (SQLException e) {
-        dataAccessor.onException(e, Read);
-        throw e;
       }
-    } else {
-      return new ArrayList<>();
+      switch (retentionPolicy) {
+        case DEFAULT_RETENTION_POLICY:
+          datasetVersionRecordList =
+              getDatasetVersionOutOfRetentionByDefault(accountId, containerId, datasetName, retentionCount,
+                  versionSchema);
+          dataAccessor.onSuccess(Read, System.currentTimeMillis() - startTimeMs);
+          return datasetVersionRecordList;
+        case ESPRESSO_BACKUP_RETENTION_POLICY:
```

Comment:
Can we avoid using ESPRESSO as the name of the retention policy? Instead, can we use the "meaning" of the policy itself? 
Is the policy means "keep single version"?

Reply:
It's not just keep single version, the detail retention policy will be implemented by espresso team, so right now I'll just rename to TIMESTAMP_BASED_BACKUP_RETENTION_POLICY.